### ISSUE TITLE
Move to a new CI server

### ISSUE THREAD

**Issue by ecampidoglio** [AutoFixture/AutoFixture#456]
I noticed that there had been [some problems](https://github.com/AutoFixture/AutoFixture/compare/381977fabdf5397458ba203f936b3b520eef0ed8...356ea3a691e487b94b85e9f5f16731cd850fcd19) with [our TC build on CodeBetter](http://teamcity.codebetter.com/project.html?projectId=AutoFixture&tab=projectOverview) lately, so I started to investigate alternative build servers, like [Travis CI](https://travis-ci.org) and [AppVeyor](https://ci.appveyor.com).

Right now, [Travis CI](https://travis-ci.org) runs .NET projects on Mono so that's not really an option for AutoFixture. [AppVeyor](https://ci.appveyor.com), on the other hand, supports Windows natively and it seems to be pretty popular among open source .NET projects.

I created an account for [AutoFixture](https://ci.appveyor.com/project/AutoFixture/autofixture) on AppVeyor but I _disabled_ the [webhook](https://github.com/AutoFixture/AutoFixture/settings/hooks/6150522) that automatically starts new builds. So far this is what I implemented:

  - [x] Build all solutions using `msbuild BuildRelease.msbuild /target:Build`
  - [x] Run all tests using `msbuild BuildRelease.msbuild /target:Test`
  - [x] Create NuGet packages using `msbuild BuildRelease.msbuild /target:NuGetPack`
  - [x] Create release archive using `msbuild BuildRelease.msbuild /target:ZipReleaseFiles`
  - [x] Publish the NuGet packages as artifacts from the build
  - [x] Push the NuGet packages to [NuGet.org](https://www.nuget.org/packages/AutoFixture/)
  - [x] Create [a GitHub release for new tags](http://www.appveyor.com/docs/deployment/github#promoting-selected-tag-to-github-release) (_optional_)

What are you thoughts on this?

**Comment by moodmosaic** [comment]
This looks promising to me! Good work, @ecampidoglio :+1:

Are we sure that [these problems](https://github.com/AutoFixture/AutoFixture/compare/381977fabdf5397458ba203f936b3b520eef0ed8...356ea3a691e487b94b85e9f5f16731cd850fcd19) - or similar - won't show up on AppVeyor?

**Comment by adamchester** [comment]
@ecampidoglio looks good :+1:

I have a few questions/comments:

 1. How would we handle _build numbers_ here, and how would they relate to NuGet/Assembly versions (if at all)? What about for builds based on branches or pull requests?
 2. AppVeyor can publish the nuget packages for every build including PRs (for example, if you wanted to test out a PR by adding a nuget source to the build, you could). Is that something we might want?
 3. I notice above you separated the different targets. I wondered myself it would be best to do _CompleteBuild_ to see if _anything_ gives an error? Then, assuming everything is OK, the NuGet packages are available for testing.
 4. Just FYI - I think AutoFixture _nearly_ builds on Mono (therefore, Travis CI could be an additional option at some point without too much effort - I guess).
 5. I wonder if publishing NuGet packages "officially" should be separate and deliberate step?

@moodmosaic 100% guaranteed! (just joking, how could we possibly be _sure_? :)

**Comment by moodmosaic** [comment]
@adamchester Right! :smiley:

(Well, I was thinking more of whether moving to AppVeyor would improve things, or not, and how.)

**Comment by ploeh** [comment]
@adamchester, we'd handle versioning in the [same way we currently do](http://blog.ploeh.dk/2013/12/10/semantic-versioning-with-continuous-deployment), unless someone has a better suggestion.

**Comment by ploeh** [comment]
Isn't there a *Verify* build step missing?

The *Verify* build configuration runs static code analysis, which isn't done by the release build, because we don't want those `[Suppress]` attributes to be visible in the released binaries.

**Comment by ecampidoglio** [comment]
The _Build_ target [depends on _Verify_](https://github.com/AutoFixture/AutoFixture/blob/master/BuildRelease.msbuild#L97), so it's always going to run that first.

I agree with @ploeh in that we should continue to version _deliberately_ – since we follow Semantic Versioning – and not let it be determined by the _build number_.

My idea was to have a build that gets triggered by commits on `master` and on any open PRs. This build would run the entire process _excluding deployment_ to NuGet.org.

The deployment would only be triggered by pushing [a new tag](http://www.appveyor.com/docs/deployment#deploy-on-tag-github-only) on `master`, which is what usually happens when we release a new version of AutoFixture. After the deployment, the build would automatically [create a new GitHub release](http://www.appveyor.com/docs/deployment/github) for the published version.

What do you think?

**Comment by ploeh** [comment]
That sounds like a fine plan :+1:

**Comment by moodmosaic** [comment]
Sounds good to me, too :+1:

I think, this is *very* cool:

>The deployment would only be triggered by pushing [a new tag](http://www.appveyor.com/docs/deployment#deploy-on-tag-github-only) on `master`, which is what usually happens when we release a new version of AutoFixture. [..]

**Comment by adamchester** [comment]
:+1:

**Comment by ecampidoglio** [comment]
The new build is now up and running :smile:

I've implemented and verified every step of the process _except_ the deployment to NuGet.org. That's tricky to verify because we don't want to push a bunch of test packages to NuGet. I'm working on it :+1:

I took the liberty to add a build status badge to the [README file](https://github.com/AutoFixture/AutoFixture/blob/master/README.md) because that's what the cool projects are doing these days. Let me know if you want to change it to something else or remove it altogether.

One thing to take into consideration is that Appveyor doesn't have an environment variable containing the commit message of an [annotated tag](https://git-scm.com/book/en/v2/Git-Basics-Tagging#Annotated-Tags). That's unfortunate, since we use that message to describe [what's in a new release](https://github.com/AutoFixture/AutoFixture/releases/tag/v3.36.9). The only piece of information available to me for the release notes of the automatically created GitHub release is the commit message of the commit _referenced by the tag_.

So, if we tag a commit _whose message_ contains the release notes instead of "_Incremented patch version_" like we use to, that commit message will automatically become the release notes for the GitHub release. I realize it's not ideal, but I think it's the best thing we can do right now. The alternative is to automatically create a release _draft_ and then manually edit the release notes before publishing it.

**Comment by ploeh** [comment]
So far, we've never created GitHub *releases* - they're all just annotated tags. Is there any reason to change that?

**Comment by ecampidoglio** [comment]
Well, I can see a couple of _minor_ reasons:

  - Since we already package all binaries in a ZIP archive (_AutoFixture.zip_) as part of the build, we could make it available for download from a GitHub release.

  - GitHub releases have a richer UI than annotated tags – for example the release notes are immediately visible and you can also see how many commits have been made since that release:

![autofixturerelease](https://cloud.githubusercontent.com/assets/877348/10817124/2c0f829a-7e38-11e5-8014-d22da91942c8.PNG)

  - It's possible to [link to the latest release](https://help.github.com/articles/linking-to-releases/#linking-to-the-latest-release) by using the same URL, for example:
`https://github.com/AutoFixture/AutoFixture/releases/latest`

None of these things are _must-haves_ by any means. I see them more like nice additions to the project that come essentially free :smiley:

**Comment by ploeh** [comment]
The more friendly we can make the release notes, the better, ceteris paribus.

Once we start introducing disadvantages to gain an advantage, though, I'd like to weigh the advantages against the disadvantages.

# Advantages

What are the advantages of nicely formatted GitHub releases?

This is difficult to answer, since we don't know how (if at all) AutoFixture users *use* the release notes.

As far as I can see, the *information* contained in a GitHub release is the same as in a tag. The only additional data associated with a release is a .zip file with the compiled binaries, but these are already available on NuGet, so will anyone ever use these?

Apart from that, the formatting of a release looks nicer than a tag, so in that sense it's friendlier.

# Disadvantages

*GitHub releases* is a proprietary GitHub feature, so I'm a little apprehensive of changing the use of Git too much only to make GitHub happy.

Granted, I already do that whenever I reference and close issues from commits, but as you may have noticed, I always make an effort of making the commit message as self-contained as possible, including also adding the URL of the issue, instead of only the issue number. As long as GitHub [URLs are cool](http://www.w3.org/Provider/Style/URI.html) (which is already a *huge* assumption), then that ought to be a fairly safe practice.

Still, the purpose of Git commits is to capture changes to source code, and commit messages should reflect that purpose. Thus, commit messages should describe what changed, or, more importantly, *why* it changed.

Commit messages should not be used to annotate releases, because there's an explicit mechanism in Git for that: *tags*.

# TL; DR

This situation reminds me a bit of the discussion associated with [GhostDoc](http://submain.com/products/ghostdoc.aspx). Many people have correctly pointed out that if the only documentation added is documentation created by an automated process, then that documentation is worthless. Any reader that understands e.g. C# will be able to infer that information from the API already.

Isn't the proposed use of GitHub releases similar to that? If we aim for a completely automated process, then what value does it add?

**Comment by ecampidoglio** [comment]
I understand your point. We shouldn't change the purpose of the commit message just to work around a limitation in the CI server.

Until AppVeyor adds an environment variable to get the message associated with an annotated tag, we could manually copy the release notes from the tag message as a last step.

**Comment by ecampidoglio** [comment]
As of [v3.36.10](https://github.com/AutoFixture/AutoFixture/releases/v3.36.10) we officially moved to a new [build on AppVeyor](https://ci.appveyor.com/project/AutoFixture/autofixture). :tada:

**Comment by ploeh** [comment]
:tada: :cake: :cocktail:

**Comment by moodmosaic** [comment]
:rocket: :beers::+1:

**Comment by adamchester** [comment]
:-1: :tada:
